Index: pisi/operations/upgrade.py
===================================================================
--- pisi/operations/upgrade.py	(revision 23095)
+++ pisi/operations/upgrade.py	(working copy)
@@ -244,15 +244,12 @@
             if extra_installs:
                 ctx.ui.warning(_('Safety switch: Following packages in system.base will be installed: ') +
                                util.strlist(extra_installs))
-            G_f, install_order = operations.install.plan_install_pkg_names(extra_installs)
-            extra_upgrades = filter(lambda x: is_upgradable(x, ignore_build), systembase - set(install_order))
-            upgrade_order = []
+            extra_upgrades = filter(lambda x: is_upgradable(x, ignore_build), systembase - set(extra_installs))
             if extra_upgrades:
                 ctx.ui.warning(_('Safety switch: Following packages in system.base will be upgraded: ') +
                                util.strlist(extra_upgrades))
-                G_f, upgrade_order = plan_upgrade(extra_upgrades)
             # return packages that must be added to any installation
-            return set(install_order + upgrade_order)
+            return set(extra_installs + extra_upgrades)
         else:
             ctx.ui.warning(_('Safety switch: the component system.base cannot be found'))
     return set()
