Index: pisi/archive.py
===================================================================
--- pisi/archive.py	(revision 16226)
+++ pisi/archive.py	(working copy)
@@ -114,8 +114,7 @@
         elif self.type == 'tarlzma':
             rmode = 'r:'
             self.file_path = self.file_path.rstrip(ctx.const.lzma_suffix)
-            ret, out, err = util.run_batch("lzma d %s %s" % (self.file_path + ctx.const.lzma_suffix,
-                                                             self.file_path))
+            ret, out, err = util.run_batch("lzma -k -f -d %s%s" % (self.file_path,ctx.const.lzma_suffix))
             if ret != 0:
                 raise LzmaRuntimeError(err)
         else:
@@ -190,9 +189,9 @@
         if self.tar.mode == 'wb' and self.type == 'tarlzma':
             batch = None
             if ctx.config.values.build.compressionlevel:
-                batch = "lzmash -%s %s" % (ctx.config.values.build.compressionlevel, self.file_path)
+                batch = "lzma -%s -z %s" % (ctx.config.values.build.compressionlevel, self.file_path)
             else:
-                batch = "lzmash %s" % self.file_path
+                batch = "lzma -z %s" % self.file_path
 
             ret, out, err = util.run_batch(batch)
             if ret != 0:
