--- lib/readline/display.c
+++ lib/readline/display.c	2004-07-30 17:02:51.000000000 +0200
@@ -1387,7 +1387,7 @@
 	      insert_some_chars (nfd, lendiff, col_lendiff);
 	      _rl_last_c_pos += col_lendiff;
 	    }
-	  else if (*ols == 0 && lendiff > 0)
+	  else if ((MB_CUR_MAX == 1 || rl_byte_oriented != 0) && (*ols == 0 && lendiff > 0))
 	    {
 	      /* At the end of a line the characters do not have to
 		 be "inserted".  They can just be placed on the screen. */
--- lib/readline/histexpand.c
+++ lib/readline/histexpand.c	2004-07-30 17:02:10.000000000 +0200
@@ -206,6 +206,7 @@
 
   /* Only a closing `?' or a newline delimit a substring search string. */
   for (local_index = i; c = string[i]; i++)
+    {
 #if defined (HANDLE_MULTIBYTE)
     if (MB_CUR_MAX > 1 && rl_byte_oriented == 0)
       {
@@ -221,8 +222,8 @@
 	    i += v - 1;
 	    continue;
 	  }
-      }
-    else
+    }
+
 #endif /* HANDLE_MULTIBYTE */
       if ((!substring_okay && (whitespace (c) || c == ':' ||
 	  (history_search_delimiter_chars && member (c, history_search_delimiter_chars)) ||
@@ -230,6 +231,7 @@
 	  string[i] == '\n' ||
 	  (substring_okay && string[i] == '?'))
 	break;
+  }
 
   which = i - local_index;
   temp = (char *)xmalloc (1 + which);
